solr/core/src/java/org/apache/solr/storage/SizeAwareDirectory.java

```diff
@@ -284,23 +365,37 @@ private SizeAccountingIndexOutput(
       this.fileSizeMap = fileSizeMap;
     }
 
-    public long setSizeWriter(LongObjectProcedure<String> sizeWriter) {
+    public Sizes setSizeWriter(SizeWriter sizeWriter) {
       if (this.sizeWriter == sizeWriter) {
-        return 0;
+        return new Sizes(0, 0);
       } else {
         // NOTE: there's an unavoidable race condition between these two
         // lines, and this ordering may occasionally overestimate directory size.
         this.sizeWriter = sizeWriter;
-        return backing.getFilePointer();
+        long onDiskSize = this.getFilePointer();
+        if (backing instanceof CompressingDirectory.SizeReportingIndexOutput) {
+          onDiskSize = ((CompressingDirectory.SizeReportingIndexOutput) backing).getBytesWritten();
+        }
```

Comment:
This is the same issue as the accounting drift, but it only happens on init, so it's harder to notice. I think we want to do the same here as we compute `onDiskSize`: 

```
if (backing instanceof SizeReportingIndexOutput)
  onDiskSize = getBytesWritten()
else if (backingDirectory instanceof OnDiskSizeDirectory)
  onDiskSize = 0
else
  onDiskSize = getFilePointer()
```